### PR TITLE
Components: refactor `CustomGradientBar` to pass `exhaustive-deps`

### PR DESCRIPTION
**Pull request by chad1008**
## What?
Updates the `CustomGradientBar` component to satisfy the `exhaustive-deps` eslint rule (specifically by updating the `ControlPoints` subcomponent)

## Why?
Part of the effort in https://github.com/WordPress/gutenberg/pull/41166 to apply `exhuastive-deps` to the Components package

## How?

**GradientPickerDropdown**
Update the `popoverProps` `useMemo` dependency array to include `gradientPickerDomRef` instead of `gradientPickerDomRef.current`.

**`useEffect` to clean event listeners on unmount**
The dependency array is currently empty, which works to trigger the cleanup only on unmount. Unfortunately`exhaustive-deps` expects to see the function being called in the dependency array, which would cause it to trigger on every render.

To avoid that, if we wrap `cleanEventListeners` in its own `useCallback` the listeners aren't able to be cleaned properly. Using a ref of the function instead allows us to exclude it from the dependency array so our `useEffect` can provide the intended cleanup, without disrupting the actual user interactions with the component.

Took me a little bit to arrive at this solution, and I'm very open to feedback if there are other/better approaches :) 

## Testing Instructions
1. From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/custom-gradient-bar`
2. Confirm that the linter returns no errors
3. Confirm e2e tests still pass

### CODE REVIEW

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (mirka) [inline packages/components/src/custom-gradient-bar/control-points.js]: In the old version, I noticed that it could've caused a subtle bug down the line where the `onStopControlPointChange` in the `cleanEventListeners` that runs on unmount is possibly a stale value. Same with the listener removal of `onMouseMove` on L146, although it looks like it would only be a problem in rare cases where the component unmounts right between a mousedown and mouseup.

In the fixed version, we can ensure that these values are always current.

**Review by mirka — APPROVED** [review]
While I do wonder whether we would've structured these functions differently (i.e. easier to machine-verify correctness) had the exhaustive-deps rule already been in place, I think this is definitely an improvement!

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: typo

```suggestion
-   `CustomGradientBar` updated to satisfy `react/exhuastive-deps` eslint rule ([#41463](https://github.com/WordPress/gutenberg/pull/41463))
```

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [inline packages/components/src/custom-gradient-bar/control-points.js:92]: [Refs should not be added to dependency](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/) arrays — I wonder if ESLint can't pick that up, since it's just an argument of the component function (and is not typed as a ref) ?

**Review by ciampo — COMMENTED** [review]
Just flagging the bug fix in #41492 — it may be worth to give priority to that PR, and make any required adjustments in this PR once it gets merged.

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [reply packages/components/src/custom-gradient-bar/control-points.js]: Similarly to https://github.com/WordPress/gutenberg/pull/41463#discussion_r887830136 — using refs in dependency arrays is a bit of an unusual pattern that can bring to unexpected behaviors ?

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (mirka) [reply packages/components/src/custom-gradient-bar/control-points.js:92]: Yeah, I wonder if eslint will pick that up once the function is typed.

But for now, I think the article agrees that it doesn't matter:

> That's because the `logRef` value _can never change_. So it actually makes no difference whether you include it or not and I guess the authors of the lint rule decided to not bother you about something that doesn't matter. Good call.

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (mirka) [reply packages/components/src/custom-gradient-bar/control-points.js]: You know what, we can actually remove this one from the dep array because eslint can tell that it's a ref 🚀

File: packages/components/CHANGELOG.md
Comment (chad1008) [reply packages/components/CHANGELOG.md]: thank you

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (chad1008) [reply packages/components/src/custom-gradient-bar/control-points.js]: whoa, nice TIL on this one, thank you both! I've pushed a commit that returns this to an empty dep array, but still using the ref in the cleanup function so `eslint` remains happy.

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (chad1008) [reply packages/components/src/custom-gradient-bar/control-points.js:92]: Fascinating, thank you again for sharing this article. It'll be interesting to see what happens when we type this one!

Reading the above article, I agree with @mirka that it should be safe (albeit unnecessary) to keep it in the array. That also prevents ESLint from throwing a warning that `gradientPickerDomRef` is missing, since it can't yet tell what it actually is.

**Comment by chad1008** [comment]
> Just flagging the bug fix in #41492 — it may be worth to give priority to that PR, and make any required adjustments in this PR once it gets merged.

Thanks for highlighting. I'll hold off on merging this and review things again after
- #41492

File: packages/components/src/custom-gradient-bar/control-points.js
Comment (ciampo) [reply packages/components/src/custom-gradient-bar/control-points.js:92]: Agreed, let's keep it for now. Hopefully, when we type this component, ESLint will flag it as an unnecessary dependency :)

**Comment by ciampo** [comment]
Just flagging that #41492 has been merged — let's rebase and give this PR a final round of reviews :)

**Comment by chad1008** [comment]
Rebased and not seeing any new/unexpected behavior. Let me know what you think @ciampo/@mirka!

**Review by mirka — APPROVED** [review]
Looks good! 🚀

**Pull request merged**
